cfgt: index entry-map by instruction location number, not the instruction itself

smowton opened the pull request:
Since function location numbers are almost always very dense, we can use a vector rather than a map to get from an instruction to its corresponding `cfgt` node. The introduced `dense_integer_mapt` for now has just enough functionality to do this, but could be extended in future to a general-purpose flat-map class a la `boost::flat_map`.

smowton committed:
Add dense_integer_mapt type

This gives a map-like interface but vector-like lookup cost, ideal for
mapping keys that have a bijection with densely packed integers, such as
the location numbers of a well-formed goto_modelt or goto_programt.

smowton commented:
On thinking again, I'll make a "give `cfgt` a public interface" PR and then base this one on that

smowton committed:
cfgt: use dense_integer_mapt to provide instruction -> cfg node mapping

This saves a lot of allocations, memory, and time when the cfg is frequently queried.

smowton commented:
This is now rebased on #5051 which does all the cleanup work that doesn't depend on the underlying data structure; the final commit that does the real work is now much briefer

peterschrammel commented on src/util/dense_integer_map.h:
```diff
@@ -0,0 +1,160 @@
+  {
+  }
+
+  template <typename Iter>
```
Some documentation for this function would be appreciated

peterschrammel commented on src/goto-programs/cfg.h:
```diff
@@ -31,6 +32,29 @@ struct cfg_base_nodet:public graph_nodet<empty_edget>, public T
+/// Functor to convert cfg nodes into dense integers, used by \ref cfg_baset.
+/// Default implementation: the identity function.
+template <class T>
+class cfg_instruction_to_dense_integert
```
This could do with some unit tests.

smowton committed:
cfgt accessors: use map type, not template instruction type

This is required because otherwise a cfgt<..., goto_programt::targett> wouldn't be able to
accept const_targett as an argument to get_node_index(...) etc, which is undesirable. In
particular its own entry_mapt::keys() would not be usable. In general this ought to be some
constified version of 'I', but since base `cfgt` is currently restricted to either targett
or const_targett then statically specifying const_targett suffices.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: af101ef272117b3d624c43f60fe35f82814c416f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/124782661
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

martin-cs commented on src/goto-instrument/dump_c.cpp:
```diff
@@ -663,6 +663,9 @@ void dump_ct::cleanup_decl(
 
+  // goto_program2codet requires valid location numbers:
+  tmp.update();
+
```
This seems like it might fix an obscure bug or two...

martin-cs reviewed (approved):
Only a very small part touches my area of responsibility and that seems fine.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: d56ce9ecf72be812025533b3b6c6bd1283e3e2e8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/129480980
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

chrisr-diffblue reviewed (approved):
Looks good, though I agree with @peterschrammel that some unit tests would be good to have.

smowton committed:
Add unit tests for dense_integer_mapt

smowton commented:
@chrisr-diffblue added, had forgotten to `git add`...

danpoe commented on src/goto-programs/cfg.h:
```diff
@@ -187,19 +187,19 @@ class cfg_baset:public grapht< cfg_base_nodet<T, I> >
   /// node indices saves a map lookup, so it can be worthwhile when you expect
   /// to repeatedly look up the same program point.
-  entryt get_node_index(const I &program_point) const
+  entryt get_node_index(const goto_programt::const_targett &program_point) const
```
Could use pass-by-value :pear:

danpoe commented on src/goto-programs/cfg.h:
```diff
@@ -187,19 +187,19 @@ class cfg_baset:public grapht< cfg_base_nodet<T, I> >
 
   /// Get the CFG graph node relating to \p program_point.
-  nodet &get_node(const I &program_point)
+  nodet &get_node(const goto_programt::const_targett &program_point)
```
:pear:

danpoe commented on src/goto-programs/cfg.h:
```diff
@@ -187,19 +187,19 @@ class cfg_baset:public grapht< cfg_base_nodet<T, I> >
 
   /// Get the CFG graph node relating to \p program_point.
-  const nodet &get_node(const I &program_point) const
+  const nodet &get_node(const goto_programt::const_targett &program_point) const
```
:pear:

danpoe commented on src/util/dense_integer_map.h:
```diff
@@ -0,0 +1,350 @@
+  // Main store: contains <key, value> pairs, where entries at positions with
+  // no corresponding key are default-initialised and entries with a
+  // corresponding key but no value set yet have the correct key but a default-
+  // initialized key. Both of these are skipped by this type's iterators.
```
initialized key -> initialized value

danpoe reviewed (approved):
LGTM

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 41585c525e2d74e3974ef3e81b46898e01d472ce).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/129671093
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 33ceb0a402588c68d022598278ea2ed8164ad8de).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/129674156

kroening commented on src/goto-programs/cfg.h:
```diff
@@ -173,12 +207,30 @@ class cfg_baset:public grapht< cfg_base_nodet<T, I> >
+    const auto &instructions = goto_program.instructions;
+    possible_keys.reserve(
+      std::distance(instructions.begin(), instructions.end()));
+    for(auto it = instructions.begin(); it != instructions.end(); ++it)
```
why `std::distance`? On a list that's linear.
`.size()` is constant since C++11.
